### PR TITLE
Adding support for recursive blocs and associativity in ARGUMENT EXTEND

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:**  feature 

We  add support for recursive blocs and associativity of `ARGUMENT EXTEND`. Example:
```
ARGUMENT EXTEND rewstrategy0
     RIGHT ASSOCIATIVITY
     PRINTED BY { pr_strategy }
  | [ "(" rewstrategy(h) ")" ] -> { h }
  | [ "subterms" rewstrategy0(h) ] -> { StratUnary (Subterms, h) }

WITH rewstrategy
    RIGHT ASSOCIATIVITY
    PRINTED BY { pr_strategy }
  | [ rewstrategy0(h) ";" rewstrategy(h') ] -> { StratBinary (Compose, h, h') }
  | [ rewstrategy0(h) ] -> { h }
 END
```

The commit on `WITH` needs polishing, but I prefer first to get feedback on the principle. When polish, it would be direct to extend it to `VERNAC ARGUMENT EXTEND`.

In practice, all `ARGUMENT EXTEND` are naturally with a right associativity. So fixing right associativity to be the default may be considered. In particular, I updated the grammars by necessity. Maybe some rules not covered by an effective example still need an explicit `RIGHT` associativity. (This is one reason defaulting to `RIGHT` associativity might be safer for compatibility.)

It is possible that recursive blocks helps the ssr plugin to avoid some `GRAMMAR EXTEND` but I prefer to let a specialist see (even if such change could also be added directly to the PR).

The change in `doc_grammar.ml` should be checked.

This PR is preparing the removal of an "optimization" of camlp5 which allows to support a right associativity even if not asked for.

- [X] Updated test-suite
- [ ] Clean the `WITH` commit and extend it to `VERNAC ARGUMENT EXTEND`
- [ ] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [ ] Entry added in the changelog

### CODE REVIEW

**Comment by jfehrle** [comment]
I don't understand how associativity is used for nonterminals without levels, so not sure about that part.

Supporting recursive definitions in ARGUMENT EXTEND could be useful.  On the other hand, it's odd that we have 2 different ways to specify productions that have different formats and feature sets.  The original Camlp5 method already supports recursive productions without any additional feature such as `WITH`.  Wouldn't it be possible to move the definition of `rewstrategy` into a GRAMMAR EXTEND and then have the ARGUMENT EXTEND provide only the extra semantic information?

Excluding GRAMMAR EXTEND, the EXTENDs seem fairly verbose.  Each only describes a single nonterminal.  Aside from the semantic info they provide, they don't have all the features of the Campl5 style productions (e.g. inline subproductions).  I think we'd be better off if we moved the semantic information elsewhere--either as annotations in the Camlp5 format or in some new file, perhaps with a build-time check that all the necessary semantic info is present.

Also, BTW, IIUC `attribute_list` in EXTEND is always interpreted as a LIST of `attribute` and can't refer to the nonterminal `attribute_list`.  And the same for `induction_clause_list`.

We might consider backporting some of the EXTEND production syntax to the Camlp5 format to make it less cluttered, such as not requiring `IDENT` everywhere, not requiring semicolons to separate tokens and putting the bind variable in parentheses.  And introduce a `KEYWORD` metasymbol to introduce keywords.

**Comment by jfehrle** [comment]
The `doc_grammar.ml` change looks OK.  I'd encourage you to do a `make doc_gram` and examine `fullGrammar` just to see it in action (look at the diffs)--and to perhaps familiarize yourself more with `doc_grammar`.

**Comment by herbelin** [comment]
@jfehrle: there are several questions in your comment.

> Wouldn't it be possible to move the definition of `rewstrategy` into a `GRAMMAR EXTEND` and then have the `ARGUMENT EXTEND` provide only the extra semantic information?

Do you mean it as a general move of `ARGUMENT EXTEND` to `GRAMMAR EXTEND` or just for `rewstrategy`? If the latter, it seems to me (and @ppedrot was [saying](https://github.com/coq/coq/pull/12717#discussion_r458424308) the them, afaicj) that we should stick to one way to do such extensions (which would include at the same time parsing, and interpretation).

>  On the other hand, it's odd that we have 2 different ways to specify productions that have different formats and feature sets

That's right. Then, it is a matter of convention of which syntax we may prefer.

If the question is why recursive entries rather than levels, it is because I found it easier (no need to invent a notation for levels). But I'd be ok to provide instead levels, we just need a notation for that.

Somehow with `WITH`, there is also a wink to the `with` of mutual inductive or recursive blocs in Gallina.

**Comment by herbelin** [comment]
> [ARGUMENT EXTEND] don't have all the features of the Campl5 style productions (e.g. inline subproductions). I think we'd be better off if we moved the semantic information elsewhere--either as annotations in the Camlp5 format or in some new file, perhaps with a build-time check that all the necessary semantic info is present.

Why not, but that may be a lot of things to do.

As a bit of history, here is how tactic extensions were done before the `ARGUMENT EXTEND` mechanism (in Coq V6.3):

In a v file:
```
(* Parsing rule *)
Grammar tactic simple_tactic :=
  CompareRule 
       [ "Compare" comarg($com1) comarg($com2)] -> 
       [(Compare $com1 $com2)].

(* Printing rule *)
Syntax tactic level 0:
  ComparePP 
       [(Compare $com1 $com2)]    -> 
       ["Compare" [1 2] $com1 [1 2] $com2].
```
In an ml file:
```
(* Wrapper extracting and interpreting actual arguments from generic syntax *)
let dyn_compare args g =
      match args with 
       [(COMMAND com1);(COMMAND com2)]  -> 
          let c1 = pf_constr_of_com g com1
          and c2 = pf_constr_of_com g com2
          in  compare c1 c2 g
     | _  ->  error "Invalid arguments for dynamic tactic"

(* Bind syntax to tactic wrapper *)
add_tactic "Compare" dyn_compare
```
In current Coq, with `ARGUMENT EXTEND`, it just becomes (in the ml file):
```
TACTIC EXTEND compare
| [ "compare" constr(c1) constr(c2) ] -> { compare c1 c2 }
END
```
And if you want to override the default value for what was called `pf_constr_of_com` at this time, this is done in `ARGUMENT EXTEND` with `INTERPRETED BY` and `GLOBALIZED BY`.

**Comment by herbelin** [comment]
> Also, BTW, IIUC `attribute_list` in EXTEND is always interpreted as a LIST of `attribute` and can't refer to the nonterminal `attribute_list`.  And the same for `induction_clause_list`.

The `attribute_list` trick is not the clearest thing we can imagine. I did it like that to be closer to the the `foo_list` names used in the grammar entries at this time, but it is neither very orthodox nor very compatible with primitive entry name already having a similar suffix (even if the suffix is supposed to commute with the action).

> We might consider backporting some of the EXTEND production syntax to the Camlp5 format to make it less cluttered, such as not requiring `IDENT` everywhere, not requiring semicolons to separate tokens and putting the bind variable in parentheses.  And introduce a `KEYWORD` metasymbol to introduce keywords.

You mean in this case backporting the Camlp5 to the `ARGUMENT`/`TACTIC` `EXTEND` syntax? Why not also. Having a unified syntax would indeed be nice.

By the way, with camlp5 in gramlib, I wonder whether it would be worth to change `Print Grammar` so that it prints entries in a style closer to `Notation` (using e.g. `at level` rather than `LEVEL`, or `constr` rather than `constr:operconstr`). We may need first to understand better a couple of ambiguities in camlp5 before, such as the confusion between `NEXT` and a `SELF` which, with associativity, behaves like a `NEXT`, and so on.

**Comment by jfehrle** [comment]
Perhaps we can chat about this on a phone call as well.  We could add this additional feature to ARGUMENT EXTEND.  I have no objection.  OTOH we have 2 ways to represent productions (GRAMMAR EXTEND and ARGUMENT EXTEND) that look a bit different and have different capabilities.  The WITH is adding something that's already supported in GRAMMAR EXTEND.  If we can get to a single mechanism that serves all our requirements that will be simpler for us use, think about and maintain, and this PR might not be needed.  I'm not planning to make such a change anytime soon but I thought it was worth mentioning.

**Comment by herbelin** [comment]
@jfehrle: we could talk on next Thursday evening 8pm Paris time?

**Comment by jfehrle** [comment]
@herbelin To talk about possible improvements to parsing?  I don't have anything new to say yet about this particular PR.  For parsing improvements, it may make more sense to wait a little longer so I have specific ideas that I've vetted somewhat.  But I could do 8PM Paris time in any case.  On August 20th, right?

**Comment by herbelin** [comment]
> To talk about possible improvements to parsing?

I meant to talk about this PR (providing recursivity vs levels, and if levels, with which syntax). But this can be a larger discussion too. Yes, on August 20th.

**Comment by herbelin** [comment]
@ppedrot: I'd like to have your opinion about this PR. Indeed, this PR is introducing new syntax, namely support for recursive grammars, which is an alternative way to using levels.

I could also support levels, e.g. with a user syntax of the following form:
```
ARGUMENT EXTEND rewstrategy
LEVEL 1
    RIGHT ASSOCIATIVITY
    PRINTED BY { pr_strategy1 }
  | [ rewstrategy(0)(h) ";" rewstrategy(0)(h') ] -> { StratBinary (Compose, h, h') }
  | [ rewstrategy(0)(h) ] -> { h }
LEVEL 0
     RIGHT ASSOCIATIVITY
     PRINTED BY { pr_strategy0 }
  | [ "(" rewstrategy(1)(h) ")" ] -> { h }
  | [ "subterms" rewstrategy(0)(h) ] -> { StratUnary (Subterms, h) }
 END
```
Q1: Do you have an opinion on providing levels vs recursion vs both of them?

Also, we talked with @jfehrle about the diversity of Coq-level user syntaxes for syntax extensions (the `Tactic Notation` syntax, the `TACTIC`/`ARGUMENT`/`VERNAC`-`EXTEND` syntax, the `Ltac2 Notation` syntax, the camlp5 syntax used by `Print Grammar`, ...).

Q2: Do you see a way to unify this?

For instance, I don't mind renouncing to the `_list`/`_opt` trick of `Tactic Notation` and of Coq's `EXTEND` as it is a bit confusing. Also, for `Print Grammar`, we now have access to the printer in gramlib, so we could adapt it to make it closer to the syntax that users see at the Coq level.

**Comment by ppedrot** [comment]
@herbelin On second thoughts I am not sure I see the interest of adding that much complexity to the `ARGUMENT EXTEND` macro. At first I thought that at least it'd help to generate a printer automatically, but I realized that it's not even the case because you still have to provide one in the `PRINTED BY` clauses.

So, essentially, my opinion on this PR depends on the following question: what is gained here compared to an explicit parsing rule performed through `GRAMMAR EXTEND`? The latter gives you all the expressive power of the campl5 API without having to invent a fresh, partial syntax for recursive entries. I know for a thing that SSR does it in a quite ugly way, but we could simply tweak the Tacentries API to make this easier...

**Comment by herbelin** [comment]
The question is about not having too many APIs and syntaxes to do the same thing. It is also a way to explicitly say that declaring an argument is not only a parsing rule, but also a printing rule, an interning rule, an interpretation rule, a substitution rule.

Or are you saying that we could keep the `ARGUMENT EXTEND` header but change the syntax for the body of `ARGUMENT EXTEND` to camlp5 syntax?

**Comment by ppedrot** [comment]
> The question is about not having too many APIs and syntaxes to do the same thing

In that respect, this PR is making things worse then?

> It is also a way to explicitly say that declaring an argument is not only [...]

There is more you can define around generic arguments, e.g. you can extend constr quotations with arbitrary stuff and the like. The set of features registered on a generic argument is open. To me, `ARGUMENT EXTEND` is just a helper to make the easy case easy. If you want to get non-trivial instances you better register your argument by hand, as done in SSR.

**Comment by herbelin** [comment]
> > The question is about not having too many APIs and syntaxes to do the same thing
> 
> In that respect, this PR is making things worse then?

Not if we consider that camlp5 is not supposed to be used by plugin developers.

> > It is also a way to explicitly say that declaring an argument is not only [...]
> 
> There is more you can define around generic arguments, e.g. you can extend constr quotations with arbitrary stuff and the like. The set of features registered on a generic argument is open. To me, `ARGUMENT EXTEND` is just a helper to make the easy case easy. If you want to get non-trivial instances you better register your argument by hand, as done in SSR.

I don't get what is your proposal. Is it to keep things as they are? `ARGUMENT EXTEND` for easy cases and `GRAMMAR EXTEND` for more complex cases?

In particular, I have a concrete question which is blocking me to continue working on notations: the `rewstrategy` argument works by chance and it would not work any more when I force camlp5 to respect priorities (which is needed to have a correct treatment of associativity of notations). So, one way or the other,`rewstrategy` has to be written in a stratified way so that correct expected priorities are statically enforced. Should I better rewrite `rewstrategy` using `GRAMMAR EXTEND` rather than extending `ARGUMENT EXTEND`?

**Comment by ppedrot** [comment]
> Not if we consider that camlp5 is not supposed to be used by plugin developers.

In the end the Tacentries API generates tokens using the campl5 production type, so at best it's a pious lie.

> I don't get what is your proposal. Is it to keep things as they are?

The current situation is not satisfactory, but I think that adding layers of complexity to reimplement the very same thing as the base API except for the use of a half-baked, quirky syntax that is constrained by some unfortunate historical design choices is not the way to go.

> In particular, I have a concrete question which is blocking me to continue working on notations

I agree this is a real problem[1], but judging from the parsing rules of that argument, it's definitely leaning on the side of full-blown DSLs. It'd be indeed more robust to write it via a `GRAMMAR EXTEND` in ssr-style, which would solve your blocker on notations. I don't remember the details, but weren't there a handful of arguments that were having the same kind of associativity issues as well?

[1] Given the guarantees enforced by the campl5 engine and its overall implementation, I technically believe that most of the Coq syntax works by accident. In a streak of cognitive dissonance, I both dream and fear of the day we switch to a parser engine that gives you more invariants.

**Comment by herbelin** [comment]
> > In particular, I have a concrete question which is blocking me to continue working on notations
> 
> I agree this is a real problem[1], but judging from the parsing rules of that argument, it's definitely leaning on the side of full-blown DSLs. It'd be indeed more robust to write it via a `GRAMMAR EXTEND` in ssr-style, which would solve your blocker on notations. 

Honestly, in term of time, it would be faster to me to have this PR merged (which is not a big deal), but I can also backtrack and study how to move `rewstrategy` to `GRAMMAR EXTEND`.

> I don't remember the details, but weren't there a handful of arguments that were having the same kind of associativity issues as well?

Not that I know.

> [1] Given the guarantees enforced by the campl5 engine and its overall implementation, I technically believe that most of the Coq syntax works by accident. In a streak of cognitive dissonance, I both dream and fear of the day we switch to a parser engine that gives you more invariants.

I recognize here your taste for "pedagogical exaggeration"! As far as I'm concerned, I'd be curious to know if other parser would be helpful for us (though this does not go in the direction of replacing ARGUMENT EXTEND with PARSING EXTEND). I'm otherwise also ok to progressively adapt camlp5 to our needs, as I know it better and better.

**Comment by jfehrle** [comment]
> Given the guarantees enforced by the campl5 engine and its overall implementation, I technically believe that most of the Coq syntax works by accident.

"works" meaning "parses correctly", right?

>  In a streak of cognitive dissonance, I both dream and fear of the day we switch to a parser engine that gives you more invariants.

We should think about ways to evolve what we have; switching to another parser engine is likely to create havoc both for developers and users.  I've been thinking about writing some code to analyze the grammar and (for example) identify productions that are disambiguated only by their order.  We may well be able to make the parser somewhat more powerful so it's less tricky to update the grammar without breaking anything.

Not quite sure what you mean by "invariants" here.

**Comment by herbelin** [comment]
@ppedrot: unless we take the decision to renounce to `ARGUMENT EXTEND`, I'd be leaning to keep `ARGUMENT EXTEND` for `rewstrategy`.

**Comment by herbelin** [comment]
I don't know how to unblock this PR. @ppedrot and @jfehrle seem to prefer `GRAMMAR EXTEND`. On my side, I find clearer to have a command which forces to declare a printer.

**Comment by JasonGross** [comment]
@coqbot ci minimize

**Comment by JasonGross** [comment]
This seems to be a nondeterministic bug in Coq's native compilation handling, let's try again.  Does Coq fail to clean up in /tmp and then error on trying to link everything in /tmp?

@coqbot CI minimize

**Comment by JasonGross** [comment]
I've worked around the issue, let's try once more, and sorry for the noise

@coqbot CI minimize

**Comment by jfehrle** [comment]
> I don't know how to unblock this PR.

And it has been dormant for 9 months.  Clearly we need some discussion.  I would like to see the parsing code become simpler and more consistent (fewer mechanisms if possible).  It seemed that this PR might add complexity and make it harder to simplify the parsing code.  Also it would help to say how this will benefit users, what problem it would solve for them.

Perhaps we (including @ppedrot in particular) could have a discussion about how we'd like to evolve the parsing code in general to make it better.  If we can agree on a plan, I'd be willing to do some of the work.  And if we can't come to an agreement then perhaps we should not use that as a reason to block this PR.

**Comment by herbelin** [comment]
> I would like to see the parsing code become simpler and more consistent (fewer mechanisms if possible). 

I'm fine with fewer parsing mechanisms if you are ready to go in this direction (and @ppedrot seemed to find on his site that `VERNAC EXTEND` is too ad hoc to be worth being extended further).

My primary concern was to work on fixing associativity in camlp5 and fixing it revealed that the `VERNAC EXTEND` for `rewstrategy` was broken and working in practice by chance. So, strictly speaking, my primary concern is to fix `rewstrategy`.

If it is clear that `VERNAC EXTEND` is going to disappear, I can temporarily rewrite `rewstrategy` with `GRAMMAR EXTEND` together with declaration of printers and interpreters.

If it is uncertain, it would be convenient for me to have this PR merged (e.g. w/o explicitly documenting the extension, and w/o advertising the extension) so that I can continue working on the gramlib associativity issue, even if the lifetime for this small extension is a couple of months. As said, my priority in the current PR is not to spend time on `VERNAC EXTEND` per se.

Based on your message, it seems useless to consider the perspective of `VERNAC EXTEND` being instead consolidated and used for many more years ahead,  but if ever it were the case, I can also spend a few time to document or amend the current PR.

Note that one point that makes me fearing that it would not be so easy to replace `VERNAC EXTEND` is that there is a cost for replacing it (would the cost be on users' shoulders, or relying on an automatic translator, etc.?).

> Also it would help to say how this will benefit users, what problem it would solve for them.

In this case, the problem is to provide a `VERNAC EXTEND` way to parse recursive expressions with several levels of precedence (but again, the PR does not intend to favor `VERNAC EXTEND` over another mechanism, just to address a current limitation of `VERNAC EXTEND`).

> And it has been dormant for 9 months.

It is also that I had little time for Coq on the last months...

> It seemed that this PR might add complexity and make it harder to simplify the parsing code. 

I believe the added complexity is here minor relatively to the whole complexity of `VERNAC EXTEND`. That should not be a worry in the perspective of eventually removing `VERNAC EXTEND`.

> Perhaps we (including @ppedrot in particular) could have a discussion about how we'd like to evolve the parsing code in general to make it better. If we can agree on a plan, I'd be willing to do some of the work. And if we can't come to an agreement then perhaps we should not use that as a reason to block this PR.

As far as I'm concerned, that would be nice!

**Comment by JasonGross** [comment]
I'm confused why this PR would change the behavior of things, but CI says it does.  @herbelin can you reproduce the issue with the minimized file?

**Comment by jfehrle** [comment]
From [here](https://github.com/coq/coq/pull/12743#issuecomment-691920797):

> @ppedrot It'd be indeed more robust to write it via a `GRAMMAR EXTEND` in ssr-style, which would solve your blocker on notations.

I don't know all the details of `ARGUMENT EXTEND` (or `VERNAC EXTEND`, for that matter).  What would the code look like to handle `rewstrategy` "ssr-style"?  And how easy would it be for non-parser experts to use the latter form?

I suppose this is an example of "ssr-style", i.e. the productions are in `GRAMMAR EXTEND` with a separate `ARGUMENT EXTEND` to give the printing info.

```
ARGUMENT EXTEND ssrterm
     PRINTED BY { pr_ssrterm }
     INTERPRETED BY { interp_ssrterm }
     GLOBALIZED BY { glob_ssrterm } SUBSTITUTED BY { subst_ssrterm }
     RAW_PRINTED BY { pr_ssrterm }
     GLOB_PRINTED BY { pr_ssrterm }
END

GRAMMAR EXTEND Gram
  GLOBAL: ssrterm;
  ssrterm: TOP [[ k = ssrtermkind; c = Pcoq.Constr.constr -> { mk_term k c } ]];
END
```

There are 161 `ARGUMENT EXTEND`s in the mlgs.  I wouldn't mind helping to convert them to another mechanism if we agree on that as a direction.

**Comment by herbelin** [comment]
> I suppose this is an example of "ssr-style", i.e. the productions are in `GRAMMAR EXTEND` with a separate `ARGUMENT EXTEND` to give the printing info.
> 
> ```
> ARGUMENT EXTEND ssrterm
>      PRINTED BY { pr_ssrterm }
>      INTERPRETED BY { interp_ssrterm }
>      GLOBALIZED BY { glob_ssrterm } SUBSTITUTED BY { subst_ssrterm }
>      RAW_PRINTED BY { pr_ssrterm }
>      GLOB_PRINTED BY { pr_ssrterm }
> END
> 
> GRAMMAR EXTEND Gram
>   GLOBAL: ssrterm;
>   ssrterm: TOP [[ k = ssrtermkind; c = Pcoq.Constr.constr -> { mk_term k c } ]];
> END
> ```

I believe so.

> There are 161 `ARGUMENT EXTEND`s in the mlgs. I wouldn't mind helping to convert them to another mechanism if we agree on that as a direction.

The question is also about the `ARGUMENT EXTEND` in the wild.

**Comment by JasonGross** [comment]
Should we also add a test of the rewstrategy grammar associativity to the test-suite?

**Comment by ppedrot** [comment]
> If it is clear that ARGUMENT EXTEND/TACTIC EXTEND/VERNAC EXTEND are going to disappear

How on earth have you reached that conclusion @herbelin ? It's a far shot to jump from the fact that the EXTEND mechanism are ad-hoc but practical to wanting to remove them altogether?

@jfehrle EXTEND statements do **much more** than parsing extensions. The fact that some parsing rules are a bit tricky to be handled by them does not mean *at all* that we can do everything with GRAMMAR EXTEND.

File: plugins/ltac/g_rewrite.mlg
Comment (ppedrot) [inline plugins/ltac/g_rewrite.mlg:90]: This code shows why this PR is nonsense: we now have a spurious generic argument registering a crapload of data that will never be used alone.

**Review by ppedrot — CHANGES_REQUESTED** [review]
Really, stop now people. I don't see how you're reaching a *folie à deux* where people indiscriminately sway from willing to extend a macro in a way that makes no sense to remove it completely *because why not*. This PR is trying to work around a single unrelated problem, which can be solved easily going the complex way. Just do this, write the easy fix and carry on. Don't bring the sledgehammer to the discussion when we have a system that works nicely for its intended domain.

In short, please close this PR.

**Comment by herbelin** [comment]
@ppedrot: sure, I could use a `GRAMMAR EXTEND`, and I could continue working on associativity without caring more about `ARGUMENT EXTEND`.

This would not solve the question that `ARGUMENT EXTEND`, if kept, is not strong enough to express a grammar with precedences (if precedences had be controllable, I'd bet that other precedences had been used for `rewstrategy`). So, the discussion is worth in my opinion.

Second, the point of @jehrle is legitimate. Why does `ARGUMENT EXTEND` use a different syntax than camlp5 (besides integrating printing, interpretation and parsing together). Personally, I prefer the syntax of `ARGUMENT EXTEND` (to the exception of the `foo_list` and `bar_opt` tricks which are precisely too tricky), but the question of not multiplying the syntaxes for grammars makes sense (disregarding compatibility issues).

Third, my analysis on the problem with associativity was not correct. It is left associative by default but right associative for productions of the form `foo := ident foo` due to the camlp5 hack.

Fourth, I'm not so fond of this `WITH` actually, and your argument about an isolated `rewstrategy0` also goes in this direction. I'm more leaning towards the standard approach of levels.

Fifth, I think that it is important to have an idea of where we are going. Closing the PR without a conclusion about where we'd like to go seems to me of no help.

Personally, I'm now leaning more in the direction of extending `ARGUMENT EXTEND` with levels, remaining open on the question of syntax (especially considering that Ltac2 - independently of its numerous qualities - adds complexity by using yet other syntax conventions, so, why not switching to Ltac2 conventions after all) but I consider important that when we invest time, it is with some confidence that it will be useful and that we are following convergent objectives.

So:
1. if I take the time to propose a variant of `ARGUMENT EXTEND` with `LEVEL`, can I count on your support?
2. if @jfehrle proposes, supported by enough users, to unify the different ways to give syntactic rules in `Tactic Notation`, `Ltac2 Notation`, `ARGUMENT EXTEND`, ... can we count on your support?
3. If not, what is your vision? Would you prefer a status quo?

**Comment by ppedrot** [comment]
> if I take the time to propose a variant of ARGUMENT EXTEND with LEVEL, can I count on your support?

That depends on how you do it. I don't really care that you bring this syntax closer to the one of GRAMMAR EXTEND *provided* that it doesn't lead to nonsensical situations. I put quite a lot of effort in ensuring that coqpp macros were statically preventing useless arguments and the like, it's not to regress on that. Also, while not actively trying to deter you from that I think this kind of endeavour is the essence of bikeshedding and moving for the sake of moving.

> if @jfehrle proposes, supported by enough users, to unify the different ways to give syntactic rules in Tactic Notation, Ltac2 Notation, ARGUMENT EXTEND, ... can we count on your support?

No. It's worse than that: I would oppose such changes. These mechanisms are very different from one another, it does not make sense to try to unify them. Furthermore, one should really consider separately dev-facing code like EXTEND macros from user-facing ones like the notation commands. We make no promise for dev-facing macros, they are just code helper preventing boilerplate. User-facing code, meanwhile, is not a macro system, it is an integral part of the Coq proof assistant and you shouldn't try to mess with it without a *very good* motivation.

> If not, what is your vision? Would you prefer a status quo?

I am still missing a motivation for these changes, when I clearly see their negative consequences, not even mentioning the fact that the people partaking in the comments of this PR have an deficient underlying mental model of what these different mechanisms are supposed to do.

**Comment by jfehrle** [comment]
IIUC, the SSR-style approach is sufficient to handle "rewstrategy", and if so, let's not add additional syntax to `ARGUMENT EXTEND`.

> if @jfehrle proposes, supported by enough users, to unify the different ways to give syntactic rules in `Tactic Notation`, `Ltac2 Notation`, `ARGUMENT EXTEND`, ... can we count on your support?

@herbelin and I discussed this informally last year.  It's a topic beyond the scope of this PR.  The next step would be to have some general discussion about how (and whether) we want to evolve the parser to make it more consistent, simpler, more powerful, etc., if reasonably possible.  On the other hand, AFAIK the current parser is not preventing us from adding features to Coq or presenting a large maintenance burden, so I turned to other tasks.

**Comment by herbelin** [comment]
> the SSR-style approach is sufficient to handle "rewstrategy",

There are different points of view. 

If we admit that `ARGUMENT EXTEND` is the relevant abstraction level for declaring the methods associated to a new element of the tactic/vernac languages, using `GRAMMAR EXTEND` would defeat this abstraction level.

Another point of view (yours, right?) is that it is overkill, or confusing, to declare the grammar rules of new elements with `ARGUMENT EXTEND` rather than with `GRAMMAR EXTEND`, considering that the latter already exists and that after all the `ARGUMENT EXTEND` grammar rules are anyway sent to gramlib.

Both points of view are ok to me, but I would feel uncomfortable, and I would find it difficult to explain to plugin developers that we deliberatedly maintain both approaches at the same time without recommending one over the other.

Also, I still don't get exactly what the position of @ppedrot is wrt to these three directions and that would be useful to know it to evaluate where we can go concretely.

**Comment by jfehrle** [comment]
> I would feel uncomfortable, and I would find it difficult to explain to plugin developers that we deliberatedly maintain both approaches at the same time without recommending one over the other.

By all means, let's add some explanation for plugin developers on which form is preferred and when the SSR-style form is necessary.  Isn't there already a document somewhere in the source tree that explains how to create plugins to which such comments could be added?

> Another point of view (yours, right?) is that it is overkill, or confusing, to declare the grammar rules of new elements with `ARGUMENT EXTEND` rather than with `GRAMMAR EXTEND`, considering that the latter already exists and that after all the `ARGUMENT EXTEND` grammar rules are anyway sent to gramlib.

My observation from quite a while back was that the metasyntax differs between the two constructs.  The `(x)`, `(c)` and `(y)` metasyntax below is not valid in `GRAMMAR EXTEND`.  For consistency, it would be nice if the metasyntax was more similar, if it's reasonably possible.  But beyond the scope of this PR.

```
ARGUMENT EXTEND test
  PRINTED BY { pr_itest' }
  INTERPRETED BY { interp_test }
  RAW_PRINTED BY { pr_test' }
  GLOB_PRINTED BY { pr_test' }
| [ int_or_var(x) comparison(c) int_or_var(y) ] -> { Test(c,x,y) }
END
```

**Comment by herbelin** [comment]
> By all means, let's add some explanation for plugin developers on which form is preferred and when the SSR-style form is necessary.

I don't have a good overview of the reasons why ssrparser uses `GRAMMAR EXTEND` on top of `ARGUMENT EXTEND`. Here are some examples:
- `ssrtacarg` and `ssrtac3arg`: I don't see the reason. Maybe to introduce a "ssr" level of abstraction to `ltac_expr` levels 3 and 5?
- `ssrsimpl_ne`, `ssrbwdview`, `ssrfwdview`, `ident_no_do`, `ssrcpat`, `ssrfwdid` : suspectingly because they use look-ahead which `ARGUMENT EXTEND` does not provide
- `ssrmmod`: suspectingly because it relies on the token `LEFTQMARK` which is not available from `ARGUMENT EXTEND`
- `ssrterm`, `ssripat`: no idea
etc.

**Comment by herbelin** [comment]
> The `(x)`, `(c)` and `(y)` metasyntax below is not valid in `GRAMMAR EXTEND`. For consistency, it would be nice if the metasyntax was more similar, if it's reasonably possible.

You mean using `;` instead of spaces to separate the symbols? But the argument could go the other way round moving the `GRAMMAR EXTEND` rules to the syntax of `ARGUMENT EXTEND` (up to a few extensions)?

**Comment by jfehrle** [comment]
> I don't have a good overview of the reasons why ssrparser uses `GRAMMAR EXTEND` on top of `ARGUMENT EXTEND`.

It may be that some (or even all) of the uses could be done without `GRAMMAR EXTEND`.  The point is that `GRAMMAR EXTEND` is necessary and sufficient to handle the `rewstrategy` case.

> You mean using `;` instead of spaces to separate the symbols? But the argument could go the other way round moving the `GRAMMAR EXTEND` rules to the syntax of `ARGUMENT EXTEND` (up to a few extensions)?

Indeed, that was my preference.  The `GRAMMAR EXTEND`s use parentheses around productions in 4 places I can see.  These might become something like `LIST0 (term LEVEL "9")(args)`, `LIST1 (LIST1 search_query) SEP "|"(l)`, etc. or we might just replace the parentheses with `[` `]` instead if that works.

```
  term:
    | "1" LEFTA
    | c = term; ".("; "@"; f = global;
        args = LIST0 (term LEVEL "9"); ")" ->
        { CAst.make ~loc @@ CAppExpl((Some (List.length args+1),f,None),args@[c]) }
  search_query:
      | b = positive_search_mark; "["; l = LIST1 (LIST1 search_query) SEP "|"; "]" -> { (b, SearchDisjConj l) }
  tactic_then_last:
    [ [ "|"; lta = LIST0 (OPT ltac_expr) SEP "|" ->
  tactic_then_last:
    [ [ "|"; lta = LIST0 (OPT ltac2_expr LEVEL "6") SEP "|" -> { lta }
```

We'd want to make sure any revised metasyntax is still yacc-parsable (which coqpp relies on).  Also we'd have to consider the impact on existing plugins that use GRAMMAR EXTEND--that might mean we support both for a release or two.  And of course, convert existing `GRAMMAR EXTEND`s to use the new form.  It seemed like a lot to take on for relatively cosmetic considerations.

Another possibility is to converge on the Ltac2 notation syntax, though it would need extensions to cover everything `GRAMMAR EXTEND` can do.

**Comment by jfehrle** [comment]
Looks like the rebate picked up extra incorrect commits.

**Comment by herbelin** [comment]
Trying #17832 as an approach based on GRAMMAR EXTEND.

**Pull request closed without merging**
